Modules/_zstd/compressor.c

```
@@ -403,6 +399,10 @@ ZstdCompressor_dealloc(PyObject *ob)
         ZSTD_freeCCtx(self->cctx);
     }
 
+    if (PyMutex_IsLocked(&self->lock)) {
+        PyMutex_Unlock(&self->lock);
+    }
```

No, no, no! This pattern will cause a fatal error if people are using the locks correctly.

---

Yeah, we should avoid this pattern. You can do something like:

```
assert(!PyMutex_IsLocked(&self->lock));
```

at the beginning of `ZstdCompressor_dealloc`, but we should structure the locking such that it's impossible for the compressor to be deallocated while the lock is held.

---

Yeah, that was what I was thinking, getting in that state should be impossible, but good to check in case it does happen somehow.

---

If we reach the destructor without releasing the lock, that's a good sign that something went horribly wrong, especially if we make sure the lock can't be held when calling into Python. I'd honestly rather a deadlock than letting it silently pass.